### PR TITLE
Test/C++-custom-filter: Add unit test cases.

### PR DESCRIPTION
**Pull request by myungjoo**
Unit test cases of C++ custom filters
    
Positive cases: 3
Negative cases: 6
    
Signed-off-by: MyungJoo Ham <myungjoo.ham@samsung.com>

### CODE REVIEW

File: tests/unittest_util.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * @brief Set pipeline state, wait until it's done.
+ * @return 0 success, -EPIPE if failed, -ETIME if timeout happens.
+ */
+extern int setPipelineStateSync (GstElement *pipeline, GstState state, uint32_t timeout_ms);
+
+/**
+ * @brief Get temp file name.
+ * @return file name (should free string with g_free)
+ */
+extern gchar * _get_temp_filename (void);
```
Comment (dongju-chae) [inline tests/unittest_util.h]: How about removing its prefix `_` because this function is now moved to the common util header (such as `getTempFilename`)

Comment (myungjoo) [reply tests/unittest_util.h]: Fixed

File: tests/cpp_methods/unittest_cpp_methods.cc [hunk 2]
```diff
@@ -15,6 +20,9 @@
 #include <nnstreamer-capi-private.h>
 
 #include <tensor_filter_cpp.h>
+#include "../unittest_util.h"
```
Comment (jaeyun-jung) [inline tests/cpp_methods/unittest_cpp_methods.cc]: Q: dependency of unittest_util includes the test directory.
```
  unittest_util_dep = declare_dependency(link_with: unittest_util_shared,
    dependencies: nnstreamer_base_deps,
    include_directories: include_directories('.')
  )
```

Comment (myungjoo) [reply tests/cpp_methods/unittest_cpp_methods.cc]: Fixed.

File: tests/cpp_methods/unittest_cpp_methods.cc [hunk 3]
```diff
@@ -48,31 +56,392 @@ class filter_basic: public tensor_filter_cpp {
...
+  char *tmp3 = getTempFilename ();
+
+  EXPECT_NE (tmp1, nullptr);
+  EXPECT_NE (tmp2, nullptr);
+  EXPECT_NE (tmp3, nullptr);
+  gchar *str_pipeline = g_strdup_printf
+      ("videotestsrc num-buffers=5 ! videoconvert ! videoscale ! video/x-raw,width=4,height=4,format=RGB ! tensor_converter ! tee name=t t. ! queue name=q1 ! tensor_filter framework=cpp model=basic_so_01,%slibcppfilter_test.so ! filesink location=%s t. ! queue name=q2 ! filesink location=%s t. ! queue ! tensor_filter framework=cpp model=basic_so_03,%slibcppfilter_test.so ! filesink location=%s", path_to_lib, tmp1, tmp2, path_to_lib, tmp3);
+
+  GError *err = NULL;
+  GstElement *pipeline = gst_parse_launch (str_pipeline, &err);
```
Comment (jaeyun-jung) [inline tests/cpp_methods/unittest_cpp_methods.cc]: clear err if failed (or set null, I think err is unnecessary in testcase)

File: tests/cpp_methods/unittest_cpp_methods.cc [hunk 4]
```diff
@@ -48,31 +56,392 @@ class filter_basic: public tensor_filter_cpp {
...
+  char *tmp2 = getTempFilename ();
+
+  EXPECT_NE (tmp1, nullptr);
+  EXPECT_NE (tmp2, nullptr);
+  EXPECT_EQ (basic._register(), 0);
+  gchar *str_pipeline = g_strdup_printf
+      ("videotestsrc num-buffers=5 ! videoconvert ! videoscale ! video/x-raw,width=4,height=4,format=RGB ! tensor_converter ! tee name=t t. ! queue name=q1 ! tensor_filter framework=cpp model=pl01 ! filesink location=%s t. ! queue name=q2 ! filesink location=%s ", tmp1, tmp2);
+
+  GError *err = NULL;
+  GstElement *pipeline = gst_parse_launch (str_pipeline, &err);
```
Comment (jaeyun-jung) [inline tests/cpp_methods/unittest_cpp_methods.cc]: clear err if failed

File: tests/cpp_methods/unittest_cpp_methods.cc [hunk 5]
```diff
@@ -48,31 +56,392 @@ class filter_basic: public tensor_filter_cpp {
...
+}
+
+/** @brief Negative case for the simpliest execution path */
+TEST (cpp_filter_on_demand, unregistered_01_n)
+{
+  filter_basic basic("basic_01");
+  gchar *str_pipeline = g_strdup_printf
+      ("videotestsrc num-buffers=3 ! videoconvert ! videoscale ! video/x-raw,width=4,height=4,format=RGB ! tensor_converter ! tensor_filter framework=cpp model=XXbasic_01 ! fakesink");
+  GError *err = NULL;
+  GstElement *pipeline = gst_parse_launch (str_pipeline, &err);
```
Comment (jaeyun-jung) [inline tests/cpp_methods/unittest_cpp_methods.cc]: clear err if failed

File: tests/cpp_methods/unittest_cpp_methods.cc [hunk 6]
```diff
@@ -48,31 +56,392 @@ class filter_basic: public tensor_filter_cpp {
...
+  char *tmp2 = getTempFilename ();
+  char *tmp3 = getTempFilename ();
+
+  EXPECT_NE (tmp1, nullptr);
+  EXPECT_NE (tmp2, nullptr);
+  EXPECT_NE (tmp3, nullptr);
+  gchar *str_pipeline = g_strdup_printf
+      ("videotestsrc num-buffers=5 ! videoconvert ! videoscale ! video/x-raw,width=4,height=4,format=RGB ! tensor_converter ! tee name=t t. ! queue name=q1 ! tensor_filter framework=cpp model=basic_so_01,%slibcppfilter_test.so ! filesink location=%s t. ! queue name=q2 ! filesink location=%s t. ! queue ! tensor_filter framework=cpp model=basic_so_03,%slibcppfilter_test.so ! filesink location=%s", path_to_lib, tmp1, tmp2, path_to_lib, tmp3);
+GError *err = NULL;
+  GstElement *pipeline = gst_parse_launch (str_pipeline, &err);
```
Comment (jaeyun-jung) [inline tests/cpp_methods/unittest_cpp_methods.cc]: clear err if failed

File: tests/cpp_methods/unittest_cpp_methods.cc [hunk 7]
```diff
@@ -48,31 +56,392 @@ class filter_basic: public tensor_filter_cpp {
...
+  EXPECT_NE (tmp1, nullptr);
+  EXPECT_NE (tmp2, nullptr);
+  EXPECT_NE (tmp3, nullptr);
+  EXPECT_NE (tmp4, nullptr);
+  EXPECT_NE (tmp5, nullptr);
+  gchar *str_pipeline = g_strdup_printf
+      ("videotestsrc num-buffers=5 ! videoconvert ! videoscale ! video/x-raw,width=4,height=4,format=RGB ! tensor_converter ! tee name=t t. ! queue name=q1 ! tensor_filter framework=cpp model=basic_so_01,%slibcppfilter_test.so ! filesink location=%s sync=true t. ! queue name=q2 ! filesink location=%s sync=true t. ! queue ! tensor_filter framework=cpp model=basic_so_02,%slibcppfilter_test.so ! filesink location=%s sync=true videotestsrc num-buffers=5 ! videoconvert ! videoscale ! video/x-raw,width=16,height=16,format=RGB ! tensor_converter ! tee name=t2 t2. ! queue ! tensor_filter framework=cpp model=basic_so2,%slibcppfilter_test.so ! filesink location=%s sync=true t2. ! queue ! filesink location=%s sync=true ", path_to_lib, tmp1, tmp2, path_to_lib, tmp3, path_to_lib, tmp4, tmp5);
+
+  GError *err = NULL;
+  GstElement *pipeline = gst_parse_launch (str_pipeline, &err);
```
Comment (jaeyun-jung) [inline tests/cpp_methods/unittest_cpp_methods.cc]: clear err if failed

File: tests/nnstreamer_sink/unittest_sink.cc [hunk 8]
```diff
@@ -16,6 +16,7 @@
 
 #include "tensor_common.h"
 #include "nnstreamer_plugin_api_filter.h"
+#include "../unittest_util.h"
```
Comment (jaeyun-jung) [inline tests/nnstreamer_sink/unittest_sink.cc]: plz fix this line also.

File: tests/cpp_methods/unittest_cpp_methods.cc (hunk 4 above)
Comment (myungjoo) [reply tests/cpp_methods/unittest_cpp_methods.cc]: Fixed.

File: tests/cpp_methods/unittest_cpp_methods.cc (hunk 5 above)
Comment (myungjoo) [reply tests/cpp_methods/unittest_cpp_methods.cc]: Fixed

File: tests/cpp_methods/unittest_cpp_methods.cc (hunk 6 above)
Comment (myungjoo) [reply tests/cpp_methods/unittest_cpp_methods.cc]: Fixed.

File: tests/cpp_methods/unittest_cpp_methods.cc (hunk 3 above)
Comment (myungjoo) [reply tests/cpp_methods/unittest_cpp_methods.cc]: Fixed.

File: tests/cpp_methods/unittest_cpp_methods.cc (hunk 7 above)
Comment (myungjoo) [reply tests/cpp_methods/unittest_cpp_methods.cc]: Fixed.

File: tests/nnstreamer_sink/unittest_sink.cc (hunk 8 above)
Comment (myungjoo) [reply tests/nnstreamer_sink/unittest_sink.cc]: Fixed.

**Pull request merged**